torch/distributed/_composable/fully_shard.py

```diff
@@ -45,6 +47,8 @@ def fully_shard(
     # Enforce the new auto wrap policy
     if policy is not None and not isinstance(policy, _FSDPPolicy):
         raise ValueError(f"Expects an `_FSDPPolicy` but got {policy}")
+    # This is a hack to make _FSDPState to be compatible with the existing contract.
```

Comment:
Can you explain why `fully_shard.set_state_module, _FSDPState())` is considered a hack to make `_FSDPState` to be compatible with the contract?

Reply:
I don't like the current code flow that `contract` first creates a `_State` instance and then `fully_shard` creates another one and overwrite the existing one. I think a better way to do that is to that `contract` wrapper to accept a `_State` subclass and instantiate the class.

Reply:
Agree, it is true for the`replicate` API as well, it would be better to allow the API to pass in a State object